datafusion/src/logical_plan/builder.rs

```diff
@@ -287,16 +287,125 @@ impl LogicalPlanBuilder {
                 .into_iter()
                 .zip(join_keys.1.into_iter())
                 .map(|(l, r)| {
-                    let mut swap = false;
                     let l = l.into();
-                    let left_key = l.clone().normalize(&self.plan).or_else(|_| {
-                        swap = true;
-                        l.normalize(right)
-                    });
-                    if swap {
-                        (r.into().normalize(&self.plan), left_key)
-                    } else {
-                        (left_key, r.into().normalize(right))
+                    let r = r.into();
+                    let lr = l.relation.clone();
+                    let rr = r.relation.clone();
+
+                    match (lr, rr) {
+                        (Some(lr), Some(rr)) => {
+                            let l_is_left =
+                                self.plan.all_schemas().iter().any(|schema| {
+                                    schema.fields().iter().any(|field| {
+                                        field.qualifier().unwrap() == &lr
+                                            && field.name() == &l.name
+                                    })
+                                });
+                            let l_is_right = right.all_schemas().iter().any(|schema| {
+                                schema.fields().iter().any(|field| {
+                                    field.qualifier().unwrap() == &lr
+                                        && field.name() == &l.name
+                                })
+                            });
+                            let r_is_left =
+                                self.plan.all_schemas().iter().any(|schema| {
+                                    schema.fields().iter().any(|field| {
+                                        field.qualifier().unwrap() == &rr
+                                            && field.name() == &r.name
+                                    })
+                                });
+                            let r_is_right = right.all_schemas().iter().any(|schema| {
+                                schema.fields().iter().any(|field| {
+                                    field.qualifier().unwrap() == &rr
+                                        && field.name() == &r.name
+                                })
+                            });
+                            match (l_is_left, l_is_right, r_is_left, r_is_right) {
+                                (true, _, _, true) => (Ok(l), Ok(r)),
+                                (_, true, true, _) => (Ok(r), Ok(l)),
+                                (_, _, _, _) => (
+                                    Err(DataFusionError::Plan(format!(
```

Comment:
couldn't this error also happen if the column was found in both the left and right input?

Reply:
Yes, I was leaving this until an initial review to understand if this was completely in the wrong direction. I can add further conditions and appropriate messages.